File: Excel/700-799/757/757 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/700-799/757/757 Alignment of Zoo Animals & Birds.xlsx"
input = read_excel(path, range = "A2:A14")
test  = read_excel(path, range = "C2:E6")

result <- input %>%
  mutate(
    zoo = if_else(str_detect(Data, "^Zoo"), Data, NA_character_),
    category = if_else(Data %in% c("Animal","Bird"), Data, NA_character_)
  ) %>%
  fill(zoo, category) %>%
  filter(!Data %in% c("Animal","Bird"), Data != zoo) %>%
  select(zoo, category, item = Data) %>%
  group_by(zoo, category) %>%
  mutate(r = row_number()) %>%
  pivot_wider(names_from = category, values_from = item) %>%
  select(Zoo = zoo, Animal, Bird)

all.equal(result, test, check.attributes = FALSE)
# > [1] TRUE

File: Excel/700-799/757/757 Challenge.py
import pandas as pd

path = "700-799/757/757 Alignment of Zoo Animals & Birds.xlsx"
input = pd.read_excel(path, usecols="A", skiprows=1, nrows=13)
test = pd.read_excel(path, usecols="C:E", skiprows=1, nrows=4)

input["zoo"] = input["Data"].where(input["Data"].str.startswith("Zoo"), pd.NA)
input["category"] = input["Data"].where(input["Data"].isin(["Animal", "Bird"]), pd.NA)
input[["zoo", "category"]] = input[["zoo", "category"]].ffill()

filtered = input[
    (~input["Data"].isin(["Animal", "Bird"])) &
    (input["Data"] != input["zoo"])
]

filtered = filtered[["zoo", "category", "Data"]].rename(columns={"Data": "item"})
filtered["r"] = filtered.groupby(["zoo", "category"]).cumcount() + 1

result = filtered.pivot(index=["zoo", "r"], columns="category", values="item").reset_index()
result = result.drop(columns="r").rename(columns={"zoo": "Zoo"})
result = result[["Zoo", "Animal", "Bird"]]

print(result.equals(test))
